File: docs/examples/rbac/go/app_test.go
package main

import (
	"testing"

	osoErrors "github.com/osohq/go-oso/errors"
)

func TestPolicy(t *testing.T) {
	oso := SetupOso()

	alphaAssociation := Organization{Name: "Alpha Association"}
	betaBusiness := Organization{Name: "Beta Business"}

	affineTypes := Repository{Name: "Affine Types", Organization: alphaAssociation}
	allocator := Repository{Name: "Allocator", Organization: alphaAssociation}
	bubbleSort := Repository{Name: "Bubble Sort", Organization: betaBusiness}
	benchmarks := Repository{Name: "Benchmarks", Organization: betaBusiness}

	ariana := NewUser("Ariana")
	bhavik := NewUser("Bhavik")

	ariana.AssignRoleForResource("owner", alphaAssociation)
	bhavik.AssignRoleForResource("contributor", bubbleSort)
	bhavik.AssignRoleForResource("maintainer", benchmarks)

	var err error
	if err = oso.Authorize(ariana, "read", affineTypes); err != nil {
		t.Errorf("Expected success; received: %v", err)
	}
	if err = oso.Authorize(ariana, "push", affineTypes); err != nil {
		t.Fatalf("Expected success; received: %v", err)
	}
	if err = oso.Authorize(ariana, "read", allocator); err != nil {
		t.Fatalf("Expected success; received: %v", err)
	}
	if err = oso.Authorize(ariana, "push", allocator); err != nil {
		t.Fatalf("Expected success; received: %v", err)
	}
	err = oso.Authorize(ariana, "read", bubbleSort)
	switch err.(type) {
	default:
		t.Fatalf("Expected NotFoundError; received: %v", err)
	case *osoErrors.NotFoundError:
	}
	err = oso.Authorize(ariana, "push", bubbleSort)
	switch err.(type) {
	default:
		t.Fatalf("ENotFoundError; received: %v", err)
	case *osoErrors.NotFoundError:
	}
	err = oso.Authorize(ariana, "read", benchmarks)
	switch err.(type) {
	default:
		t.Fatalf("ENotFoundError; received: %v", err)
	case *osoErrors.NotFoundError:
	}
	err = oso.Authorize(ariana, "push", benchmarks)
	switch err.(type) {
	default:
		t.Fatalf("ENotFoundError; received: %v", err)
	case *osoErrors.NotFoundError:
	}

	err = oso.Authorize(bhavik, "read", affineTypes)
	switch err.(type) {
	default:
		t.Fatalf("ENotFoundError; received: %v", err)
	case *osoErrors.NotFoundError:
	}
	err = oso.Authorize(bhavik, "push", affineTypes)
	switch err.(type) {
	default:
		t.Fatalf("ENotFoundError; received: %v", err)
	case *osoErrors.NotFoundError:
	}
	err = oso.Authorize(bhavik, "read", allocator)
	switch err.(type) {
	default:
		t.Fatalf("ENotFoundError; received: %v", err)
	case *osoErrors.NotFoundError:
	}
	err = oso.Authorize(bhavik, "push", allocator)
	switch err.(type) {
	default:
		t.Fatalf("ENotFoundError; received: %v", err)
	case *osoErrors.NotFoundError:
	}
	if err = oso.Authorize(bhavik, "read", bubbleSort); err != nil {
		t.Fatalf("Expected success; received: %v", err)
	}
	err = oso.Authorize(bhavik, "push", bubbleSort)
	switch err.(type) {
	default:
		t.Fatalf("Expected ForbiddenError; received: %v", err)
	case *osoErrors.ForbiddenError:
	}
	if err = oso.Authorize(bhavik, "read", benchmarks); err != nil {
		t.Fatalf("Expected success; received: %v", err)
	}
	if err = oso.Authorize(bhavik, "push", benchmarks); err != nil {
		t.Fatalf("Expected success; received: %v", err)
	}
}

File: docs/examples/rbac/nodejs/app.test.js
const { Oso } = require('oso');

const { Organization, Repository, Role, User } = require('./app');

const oso = new Oso();

oso.registerClass(Organization);
oso.registerClass(Repository);
oso.registerClass(User);

beforeEach(() => oso.loadFiles(["main.polar"]));
afterEach(() => oso.clearRules());

const alphaAssociation = new Organization("Alpha Association");
const betaBusiness = new Organization("Beta Business");

const affineTypes = new Repository("Affine Types", alphaAssociation);
const allocator = new Repository("Allocator", alphaAssociation);
const bubbleSort = new Repository("Bubble Sort", betaBusiness);
const benchmarks = new Repository("Benchmarks", betaBusiness);

const ariana = new User("Ariana");
const bhavik = new User("Bhavik");

ariana.assignRoleForResource("owner", alphaAssociation);
bhavik.assignRoleForResource("contributor", bubbleSort);
bhavik.assignRoleForResource("maintainer", benchmarks);

test('policy', async () => {
    await expect(oso.authorize(ariana, "read", affineTypes)).resolves.toBeUndefined();
    await expect(oso.authorize(ariana, "push", affineTypes)).resolves.toBeUndefined();
    await expect(oso.authorize(ariana, "read", allocator)).resolves.toBeUndefined();
    await expect(oso.authorize(ariana, "push", allocator)).resolves.toBeUndefined();
    await expect(oso.authorize(ariana, "read", bubbleSort)).rejects.toThrow('404');
    await expect(oso.authorize(ariana, "push", bubbleSort)).rejects.toThrow('404');
    await expect(oso.authorize(ariana, "read", benchmarks)).rejects.toThrow('404');
    await expect(oso.authorize(ariana, "push", benchmarks)).rejects.toThrow('404');

    await expect(oso.authorize(bhavik, "read", affineTypes)).rejects.toThrow('404');
    await expect(oso.authorize(bhavik, "push", affineTypes)).rejects.toThrow('404');
    await expect(oso.authorize(bhavik, "read", allocator)).rejects.toThrow('404');
    await expect(oso.authorize(bhavik, "push", allocator)).rejects.toThrow('404');
    await expect(oso.authorize(bhavik, "read", bubbleSort)).resolves.toBeUndefined();
    await expect(oso.authorize(bhavik, "push", bubbleSort)).rejects.toThrow('403');
    await expect(oso.authorize(bhavik, "read", benchmarks)).resolves.toBeUndefined();
    await expect(oso.authorize(bhavik, "push", benchmarks)).resolves.toBeUndefined();
});
